File: apps/macos/Sources/CueWeaveMac/ProjectModels.swift
import Foundation
import UniformTypeIdentifiers

extension UTType {
    static let cueWeaveProject = UTType(exportedAs: "dev.cueweave.project", conformingTo: .json)
}

func resolveProjectPath(_ path: String, relativeTo projectURL: URL?) -> URL {
    let portable = path.replacingOccurrences(of: "\\", with: "/")
    let foreignAbsolute = portable.hasPrefix("/") || portable.dropFirst().first == ":"
    guard !foreignAbsolute, let projectURL else { return URL(fileURLWithPath: path) }
    return projectURL.deletingLastPathComponent().appendingPathComponent(portable)
}

struct ProjectDocument: Codable, Equatable {
    var schemaVersion: Int
    var source: SourceInfo?
    var target: TargetAudio?
    var metadata: MetadataSet
    var lyrics: LyricsDocument
    var timeline: [Cue]
    var exportProfile: ExportProfile

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case source, target, metadata, lyrics, timeline
        case exportProfile = "export"
    }
}

struct SourceInfo: Codable, Equatable {
    var path: String
    var fingerprint: MediaFingerprint?
    var musicID: UInt64?
    var coverURL: String?
    var format: String?
    var durationMS: UInt64?

    enum CodingKeys: String, CodingKey {
        case path, fingerprint, format
        case musicID = "music_id"
        case coverURL = "cover_url"
        case durationMS = "duration_ms"
    }
}

struct TargetAudio: Codable, Equatable {
    var path: String
    var fingerprint: MediaFingerprint?
    var durationMS: UInt64?

    enum CodingKeys: String, CodingKey {
        case path, fingerprint
        case durationMS = "duration_ms"
    }
}

struct MediaFingerprint: Codable, Equatable {
    var fileName: String
    var sizeBytes: UInt64
    var sha256: String

    enum CodingKeys: String, CodingKey {
        case fileName = "file_name"
        case sizeBytes = "size_bytes"
        case sha256
    }
}

struct MetadataSet: Codable, Equatable {
    var source: MetadataValues
    var target: MetadataValues
    var draft: MetadataValues
}

struct MetadataValues: Codable, Equatable {
    var title: String?
    var artists: [String]
    var albumArtist: String?
    var album: String?
    var track: UInt32?
    var disc: UInt32?
    var date: String?
    var composer: String?
    var lyricist: String?
    var coverPath: String?

    enum CodingKeys: String, CodingKey {
        case title, artists, album, track, disc, date, composer, lyricist
        case albumArtist = "album_artist"
        case coverPath = "cover_path"
    }
}

struct LyricsDocument: Codable, Equatable {
    var credits: [Credit]
    var lines: [LyricLine]
}

struct Credit: Codable, Equatable {
    var label: String
    var value: String
}

struct LyricLine: Codable, Equatable, Identifiable {
    var id: UInt64
    var original: String
    var translation: String?
    var segments: [LyricSegment]
}

struct LyricSegment: Codable, Equatable, Identifiable {
    var id: UInt64
    var text: String
    var timing: SegmentTiming
}

struct SegmentTiming: Codable, Equatable {
    var gemini: AlignmentPoint?
    var finalPoint: AlignmentPoint?
    var review: ReviewState

    enum CodingKeys: String, CodingKey {
        case gemini, review
        case finalPoint = "final"
    }
}

struct AlignmentPoint: Codable, Equatable {
    var timeMS: UInt64
    var confidence: Float?

    enum CodingKeys: String, CodingKey {
        case timeMS = "time_ms"
        case confidence
    }
}

enum ReviewState: String, Codable, CaseIterable {
    case pending
    case autoAccepted = "auto_accepted"
    case needsReview = "needs_review"
    case userConfirmed = "user_confirmed"
    case ignored
    case unmatched

    var title: String {
        switch self {
        case .pending: L10n.shared.t("review.pending")
        case .autoAccepted: L10n.shared.t("review.auto")
        case .needsReview: L10n.shared.t("review.review")
        case .userConfirmed: L10n.shared.t("review.confirmed")
        case .ignored: L10n.shared.t("review.ignored")
        case .unmatched: L10n.shared.t("review.unmatched")
        }
    }
}

struct Cue: Codable, Equatable {
    var type: String
    var timeMS: UInt64?
    var text: String?
    var lineID: UInt64?

    enum CodingKeys: String, CodingKey {
        case type, text
        case timeMS = "time_ms"
        case lineID = "line_id"
    }
}

struct ExportProfile: Codable, Equatable {
    var offsetMS: Int64
    var formats: [ExportFormat]
    var bilingual: BilingualMode

    enum CodingKeys: String, CodingKey {
        case offsetMS = "offset_ms"
        case formats, bilingual
    }
}

enum ExportFormat: String, Codable, CaseIterable, Identifiable {
    case lrc, uslt, sylt
    var id: String { rawValue }
    var title: String { rawValue.uppercased() }
    var detail: String {
        switch self {
        case .lrc: L10n.shared.t("export.lrc")
        case .uslt: L10n.shared.t("export.uslt")
        case .sylt: L10n.shared.t("export.sylt")
        }
    }
}

enum BilingualMode: String, Codable, CaseIterable, Identifiable {
    case originalOnly = "original_only"
    case combined
    var id: String { rawValue }
    var title: String { self == .originalOnly ? L10n.shared.t("export.originalOnly") : L10n.shared.t("export.combined") }
}

enum WorkspacePage: String, CaseIterable, Identifiable {
    case source = "Source"
    case metadata = "Metadata"
    case lyrics = "Lyrics"
    case translation = "Translation"
    case alignment = "Alignment"
    case export = "Export"

    var id: String { rawValue }
    var title: String {
        switch self {
        case .source: L10n.shared.t("page.source")
        case .metadata: L10n.shared.t("page.metadata")
        case .lyrics: L10n.shared.t("page.lyrics")
        case .translation: L10n.shared.t("page.translation")
        case .alignment: L10n.shared.t("page.alignment")
        case .export: L10n.shared.t("page.export")
        }
    }
    var icon: String {
        switch self {
        case .source: "square.and.arrow.down"
        case .metadata: "tag"
        case .lyrics: "text.quote"
        case .translation: "translate"
        case .alignment: "waveform.path"
        case .export: "square.and.arrow.up"
        }
    }

    var detail: String {
        switch self {
        case .source: L10n.shared.t("page.source.detail")
        case .metadata: L10n.shared.t("page.metadata.detail")
        case .lyrics: L10n.shared.t("page.lyrics.detail")
        case .translation: L10n.shared.t("page.translation.detail")
        case .alignment: L10n.shared.t("page.alignment.detail")
        case .export: L10n.shared.t("page.export.detail")
        }
    }
}

enum AlignmentProvider: String, CaseIterable, Identifiable {
    case openRouter = "openrouter"
    case aiStudio = "ai_studio"

    var id: String { rawValue }
    var title: String { self == .openRouter ? "OpenRouter" : "AI Studio" }
}

enum WorkspaceStageState: Equatable {
    case ready
    case review(Int)
    case pending

    var label: String {
        switch self {
        case .ready: L10n.shared.t("stage.ready")
        case let .review(count): L10n.shared.t("stage.review", String(count))
        case .pending: L10n.shared.t("stage.pending")
        }
    }
}

enum MetadataOrigin {
    case source
    case target
}

File: apps/windows/CueWeave.Windows/Models/ProjectModels.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CueWeave.WinUI.Models;

public sealed class ProjectDocument
{
    public uint SchemaVersion { get; set; }
    public SourceInfo? Source { get; set; }
    public TargetAudio? Target { get; set; }
    public MetadataSet Metadata { get; set; } = new();
    public LyricsDocument Lyrics { get; set; } = new();
    public List<JsonElement> Timeline { get; set; } = [];
    [JsonPropertyName("export")] public ExportProfile ExportProfile { get; set; } = new();

    [JsonIgnore] public List<LyricSegment> Segments => Lyrics.Lines.SelectMany(line => line.Segments).ToList();
}

public sealed class SourceInfo
{
    public string Path { get; set; } = "";
    public MediaFingerprint? Fingerprint { get; set; }
    public ulong? MusicId { get; set; }
    public string? CoverUrl { get; set; }
    public string? Format { get; set; }
    public ulong? DurationMs { get; set; }
}

public sealed class TargetAudio
{
    public string Path { get; set; } = "";
    public MediaFingerprint? Fingerprint { get; set; }
    public ulong? DurationMs { get; set; }
}

public sealed class MediaFingerprint
{
    public string FileName { get; set; } = "";
    public ulong SizeBytes { get; set; }
    public string Sha256 { get; set; } = "";
}

public sealed class MetadataSet
{
    public MetadataValues Source { get; set; } = new();
    public MetadataValues Target { get; set; } = new();
    public MetadataValues Draft { get; set; } = new();
}

public sealed class MetadataValues
{
    public string? Title { get; set; }
    public List<string> Artists { get; set; } = [];
    public string? AlbumArtist { get; set; }
    public string? Album { get; set; }
    public uint? Track { get; set; }
    public uint? Disc { get; set; }
    public string? Date { get; set; }
    public string? Composer { get; set; }
    public string? Lyricist { get; set; }
    public string? CoverPath { get; set; }
}

public sealed class LyricsDocument
{
    public List<Credit> Credits { get; set; } = [];
    public List<LyricLine> Lines { get; set; } = [];
}

public sealed class Credit
{
    public ulong Id { get; set; }
    public string Label { get; set; } = "";
    public string Value { get; set; } = "";

    [JsonIgnore]
    public string DisplayText
    {
        get
        {
            var label = Label.Trim();
            var value = Value.Trim();
            return string.IsNullOrEmpty(label) ? value : $"{label}：{value}";
        }
    }
}

public sealed class LyricLine
{
    public ulong Id { get; set; }
    public string Original { get; set; } = "";
    public string? Translation { get; set; }
    public List<LyricSegment> Segments { get; set; } = [];
}

public sealed class LyricSegment
{
    public ulong Id { get; set; }
    public string Text { get; set; } = "";
    public SegmentTiming Timing { get; set; } = new();
}

public sealed class SegmentTiming
{
    public AlignmentPoint? Gemini { get; set; }
    [JsonPropertyName("final")] public AlignmentPoint? Final { get; set; }
}

public sealed class AlignmentPoint
{
    public ulong TimeMs { get; set; }
    public float? Confidence { get; set; }
}

public sealed class ExportProfile
{
    public long OffsetMs { get; set; }
    public List<string> Formats { get; set; } = ["lrc"];
    public string Bilingual { get; set; } = "original_only";
}

[JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.SnakeCaseLower,
    WriteIndented = true, GenerationMode = JsonSourceGenerationMode.Metadata)]
[JsonSerializable(typeof(ProjectDocument))]
[JsonSerializable(typeof(LocalSettings))]
internal partial class CueJsonContext : JsonSerializerContext;
